### PR TITLE
feat(codecv1): add commit gas and calldata size estimation

### PR DESCRIPTION
**Pull request by colinlyguo**
### Purpose or design rationale of this PR

1. add codecv1 commit calldata size and gas estimations.
2. add calldata size and gas limits in chunk & batch-proposers.
3. update l2geth dependency to reduce blob copy ops.
4. more accurate blob usage for better analysis.

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] feat: A new feature

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change

### CODE REVIEW

File: common/types/encoding/codecv1/codecv1.go [hunk 1]
```diff
@@ -506,3 +504,134 @@ func chunkL1CommitBlobDataSize(c *encoding.Chunk) (uint64, error) {
...
+		blockL1CommitGas := EstimateBlockL1CommitGas(block)
+		totalL1CommitGas += blockL1CommitGas
+	}
+
+	numBlocks := uint64(len(c.Blocks))
+	totalL1CommitGas += 100 * numBlocks                         // numBlocks times warm sload
+	totalL1CommitGas += CalldataNonZeroByteGas                  // numBlocks field of chunk encoding in calldata
+	totalL1CommitGas += CalldataNonZeroByteGas * numBlocks * 60 // numBlocks of BlockContext in chunk
+
+	totalL1CommitGas += GetKeccak256Gas(58*numBlocks + 32*totalTxNum) // chunk hash
```
Comment (zimpha) [inline common/types/encoding/codecv1/codecv1.go]: only num of l1 txs are used in chunk hash.

File: rollup/internal/controller/sender/sender.go [hunk 2]
```diff
@@ -618,13 +618,13 @@ func makeSidecar(blob *kzg4844.Blob) (*gethTypes.BlobTxSidecar, error) {
 	var commitments []kzg4844.Commitment
 	var proofs []kzg4844.Proof
 
-	for _, b := range blobs {
-		c, err := kzg4844.BlobToCommitment(b)
+	for i := range blobs {
+		c, err := kzg4844.BlobToCommitment(&blobs[i])
```
Comment (georgehao) [inline rollup/internal/controller/sender/sender.go:622]: isn't it equivalent to the previous?

Comment (colinlyguo) [reply rollup/internal/controller/sender/sender.go:622]: the previous implementation would get golint error as follows:
```
internal/controller/sender/sender.go:622:38: G601: Implicit memory aliasing in for loop. (gosec)
                c, err := kzg4844.BlobToCommitment(&b)
                                                   ^
internal/controller/sender/sender.go:627:38: G601: Implicit memory aliasing in for loop. (gosec)
                p, err := kzg4844.ComputeBlobProof(&b, c)
                                                   ^
```
so changed to the current implementation.

**Comment by colinlyguo** [comment]
L1 Pos devnet is broken, causing CI failure.

File: common/types/encoding/codecv1/codecv1.go (hunk 1 above)
Comment (colinlyguo) [reply common/types/encoding/codecv1/codecv1.go]: Doesn't chunk data hash include the first 58 bytes of each block context here: https://github.com/scroll-tech/scroll/blob/v4.4.2/contracts/src/L1/rollup/ScrollChain.sol#L854?

Comment (zimpha) [reply common/types/encoding/codecv1/codecv1.go]: Sorry, I mean `totalTxNum` is not correct. Only L1 txs is used.

Comment (colinlyguo) [reply common/types/encoding/codecv1/codecv1.go]: oh, I c and nice catch. I misunderstood your comment, will fix it.

Comment (Thegaram) [reply common/types/encoding/codecv1/codecv1.go]: https://github.com/scroll-tech/scroll/blob/develop/contracts/src/L1/rollup/ScrollChain.sol#L862

Comment (colinlyguo) [reply common/types/encoding/codecv1/codecv1.go]: fixed in [6b44193](https://github.com/scroll-tech/scroll/pull/1324/commits/6b441938a04385175d1d42282c550df9edcf193f).

**Review by Thegaram — APPROVED** [review]

**Review by zimpha — APPROVED** [review]

**Comment by colinlyguo** [comment]
let's wait until the ci has passed.

**Pull request merged**